File: src/main/java/org/yanning/gradle/vcs_lib/vcsLib.java
package org.yanning.gradle.vcs_lib;

import org.gradle.api.*;
import org.yanning.gradle.vcs_lib.extension.Config;
import org.yanning.gradle.vcs_lib.task.*;

public class vcsLib implements Plugin<Project> {
    public static final String TASK_vcsLibInit = "vcsLibsInit";
    public static final String TASK_vcsLib = "vcsLibs";
    public static final String TASK_vcsLibsDemo = "vcsLibsDemo";
    public static final String TASK_vcsLibsBuild = "vcsLibsBuild";
    public static final String TASK_vcsLibsUpdate = "vcsLibsUpdate";
    public static final String TASK_vcsLibsUpload = "vcsLibsUpload";

    public static void main(String... strings) {
        System.out.print("ok");
    }

    @Override
    public void apply(Project target) {
//        target.getExtensions().add(EXT_vcsLib, config = new Config());
        target.getTasksByName("assemble", false).iterator().next().doLast(task ->
                System.err.println("success....................."));
        target.getTasks().create(TASK_vcsLib, MAIN.class, main -> {
            main.setUrl("svn//127.0.0.1/demo");

        });
        target.getTasks().create(TASK_vcsLibsDemo, Demo.class);
        target.getTasks().create(TASK_vcsLibsBuild, Build.class);
        target.getTasks().create(TASK_vcsLibsUpdate, Update.class);
        target.getTasks().create(TASK_vcsLibsUpload, Upload.class);
    }
}

File: src/main/kotlin/org/yanning/gradle/vcs_lib/vcsLib.kt
package org.yanning.gradle.vcs_lib

import org.gradle.api.Plugin
import org.gradle.api.Project
import org.yanning.gradle.vcs_lib.core.AppConfig
import org.yanning.gradle.vcs_lib.core.ConfKey
import org.yanning.gradle.vcs_lib.extension.RepositoryBuilder
import org.yanning.gradle.vcs_lib.task.GUI
import org.yanning.gradle.vcs_lib.utils.Log

class vcsLib : Plugin<Project> {

    override fun apply(target: Project) {
        val appConfig = AppConfig(target)
        val config = appConfig.conf
        if (config.getConf(ConfKey.repoUri).isEmpty()) return
        val repo = RepositoryBuilder(config).build() ?: return
        target.tasks.create(TaskConf.vcsLibUpdate.name) {
            if (repo.hasCheckout()) {
                Log.out("repo has updated,you can update repo forced use to 'gradle vcsLibUpdate'!")
            } else {
                Log.out("repo update start...")
                repo.update()
                Log.out("repo update end...")
            }
        }
        /**
         * 添加仓库地址
         */
        target.repositories.add(
                target.repositories.maven {
                    it.name = "vcsLib-${target.name}"
                    it.url = repo.outDir().toURI()
                })

        if (target.tasks.findByName(TaskConf.vcsLibGUI.name) == null) {
            /**
             * 创建GUI命令
             */
            target.tasks.create(TaskConf.vcsLibGUI.name, GUI::class.java)
        }
    }
}